### PR TITLE
Review

### PR DESCRIPTION
**Pull request by bl0rch1d**
**Heroku link:**

https://bookstore-diz.herokuapp.com/

**Coupon:**
code - **FUYWXK**
discount - 52%

### CODE REVIEW

File: app/models/author.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class Author < ApplicationRecord
+  has_and_belongs_to_many :books
+
+  validates :first_name, :last_name, presence: true
+  validates :first_name, uniqueness: { scope: :last_name }
+
+  def to_s
```
Comment (yuriy1991) [inline app/models/author.rb]: you can use draper gem to decorate the model and add `full_name` to it

File: app/admin/orders.rb [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+  index do
+    selectable_column
+
+    column :number
+    column :created_at
+    column :state
+
+    column do |resource|
+      link_to 'Change State', resource_path(resource)
```
Comment (yuriy1991) [inline app/admin/orders.rb]: i18n

File: app/models/address.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
+class Address < ApplicationRecord
```
Comment (yuriy1991) [inline app/models/address.rb]: what about billing_address? STI?

File: app/models/material.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+class Material < ApplicationRecord
+  has_and_belongs_to_many :books
+
+  validates :title, presence: true, uniqueness: true, length: { maximum: 50 }
```
Comment (yuriy1991) [inline app/models/material.rb]: 50 const

File: app/models/coupon.rb [hunk 5]
```diff
@@ -0,0 +1,10 @@
+class Coupon < ApplicationRecord
+  # belongs_to :order, optional: true
+
+  validates :code, :discount, :expire_date, presence: true
+  validates :discount, numericality: { only_integer: true, less_than: 100, greater_than: 1 }
+
+  def format_discount
```
Comment (yuriy1991) [inline app/models/coupon.rb]: this is not model logic

File: app/models/book_image.rb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+class BookImage < ApplicationRecord
+  mount_uploader :image, BookImageUploader
```
Comment (yuriy1991) [inline app/models/book_image.rb]: https://edgeguides.rubyonrails.org/active_storage_overview.html ?

File: test/controllers/home_controller_test.rb [hunk 7]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
```
Comment (yuriy1991) [inline test/controllers/home_controller_test.rb]: Do not write controller tests. Write feature tests instead

File: test/fixtures/addresses.yml [hunk 8]
```diff
@@ -0,0 +1,23 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
+one:
```
Comment (yuriy1991) [inline test/fixtures/addresses.yml]: Do not use fixtures, use factories instead:
https://semaphoreci.com/blog/2014/01/14/rails-testing-antipatterns-fixtures-and-factories.html

File: test/models/address_test.rb [hunk 9]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class AddressTest < ActiveSupport::TestCase
+  # test "the truth" do
```
Comment (yuriy1991) [inline test/models/address_test.rb]: shoulda matchers

File: .overcommit.yml [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+  Rubocop:
+    enabled: true
+
+  Fasterer:
+    enabled: true
+
+  BundleAudit:
+    enabled: true
+
+  # RailsBestPractices:
```
Comment (yuriy1991) [inline .overcommit.yml]: please turn it on

File: app/admin/books.rb [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+    active_admin_comments
+  end
+
+  form do |f|
+    f.inputs do
+      f.input :title
+      f.input :category, as: :radio
+
+      f.input :authors, as: :check_boxes, collection: Author.all.map { |author| author.decorate.full_name }
```
Comment (yuriy1991) [inline app/admin/books.rb]: you can decorate collection https://github.com/drapergem/draper#collections

File: app/assets/javascripts/book.js [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+let decrementQuantity = () => {
+  let input = document.getElementById("quantity_input");
+  
+  if (input.value > 1) { input.value--; }
+}
+
+document.addEventListener("click", (event) => {
+  console.log(event.target)
```
Comment (yuriy1991) [inline app/assets/javascripts/book.js]: remove

File: app/concepts/address/operation/update.rb [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+  def extract_params(ctx, params:, **)
+    ctx['billing_params'] = params.dig(:user, :billing_address_attributes)
+    ctx['shipping_params'] = params.dig(:user, :shipping_address_attributes)
+  end
+
+  def validate(ctx, **)
+    billing = ctx['billing_address_form'].validate(ctx['billing_params']) if ctx['billing_params']
+    shipping = ctx['shipping_address_form'].validate(ctx['shipping_params']) if ctx['shipping_params']
+
+    billing || shipping
```
Comment (yuriy1991) [inline app/concepts/address/operation/update.rb]: what if shipping is invalid?
maybe you should do smth like this?: `[billing, shipping].compact.all?(&:valid?)`

File: app/concepts/book/contract/index.rb [hunk 14]
```diff
@@ -0,0 +1,12 @@
+module Book::Contract
+  class Index < Reform::Form
+    MIN_PAGE_VALUE = 1
+    MIN_CATEGORY_ID_VALUE = 0
```
Comment (yuriy1991) [inline app/concepts/book/contract/index.rb]: where is it used?

File: app/concepts/book/operation/index.rb [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+      items: Book::PAGINATION_INDEX
+    )
+  end
+
+  def overflow?(_ctx, pagy:, **)
+    !pagy.overflow?
+  end
+
+  def overflow(ctx, **)
+    ctx['contract.default'].errors.add(:page, 'is out of limits')
```
Comment (yuriy1991) [inline app/concepts/book/operation/index.rb]: please add I18n

File: app/concepts/book/query/index.rb [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+      when I18n.t('sortings.system.title_descending') then title_descending(from_category)
+      else title_ascending(from_category)
+      end
+    end
+    # rubocop: enable Metrics/AbcSize, CyclomaticComplexity
+
+    private
+
+    def from_category
+      category_books = (@params[:category_id] ? Category.find(@params[:category_id]).books : Book.all)
```
Comment (yuriy1991) [inline app/concepts/book/query/index.rb]: you pass from_category to every method
Le'ts add `initialize`, where @books (Category.find(@params[:category_id]).books : Book.all) will be defined
What if category_id is invalid? Maybe we should do .find_by(id: id) instead of find?

File: app/concepts/checkout/operation/complete.rb [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+  def model(ctx, params:, **)
+    ctx['model'] = params[:current_order]
+  end
+
+  def clear_session(_ctx, params:, **)
+    params[:session][:current_order_id] = nil
+  end
+
+  def mail(_ctx, params:, **)
+    CheckoutMailer.complete(params[:current_user], params[:current_order]).deliver_later
```
Comment (yuriy1991) [inline app/concepts/checkout/operation/complete.rb]: it's better to pass current_used and current_order ids
https://github.com/mperham/sidekiq/wiki/Best-Practices#1-make-your-job-parameters-small-and-simple

File: app/concepts/coupon/operation/apply.rb [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+  step :model, fail_fast: true
+  step :expired?
+  step :link_to_order
+
+  def model(ctx, params:, **)
+    ctx['model'] = Coupon.find_by(code: params[:code])
+  end
+
+  def expired?(ctx, **)
+    ctx['coupon.relevant'] = Time.now.to_i <= ctx['model'].expire_date.to_i
```
Comment (yuriy1991) [inline app/concepts/coupon/operation/apply.rb]: Time.zone.now (enable rails best practices)

File: app/concepts/coupon/policy/apply.rb [hunk 19]
```diff
@@ -0,0 +1,7 @@
+class Coupon::Policy::ApplyGuard
+  include Uber::Callable
+
+  def call(_ctx, params:, **)
+    params[:order].present? && params[:order].is_a?(Order)
```
Comment (yuriy1991) [inline app/concepts/coupon/policy/apply.rb]: do we need to check order class?
I think you should check if the order belongs to the current_user

File: app/concepts/order/policy/show.rb [hunk 20]
```diff
@@ -0,0 +1,8 @@
+class Order::Policy::ShowGuard
+  include Uber::Callable
+
+  def call(_ctx, params:, **)
+    params[:current_user]&.id == params[:user_id].to_i &&
```
Comment (yuriy1991) [inline app/concepts/order/policy/show.rb]: let's extract logic to methods:
`first_condition? && second_condition?`

File: app/controllers/application_controller.rb [hunk 21]
```diff
@@ -1,2 +1,49 @@
...
+  end
+
+  def obtain_order
+    order = Order.find_or_create_by(id: session[:current_order_id])
+    session[:current_order_id] = order.id
+
+    order.decorate
+  end
+
+  def expose_address_forms(result)
```
Comment (yuriy1991) [inline app/controllers/application_controller.rb]: should it be in ApplicationController?

File: app/controllers/application_controller.rb [hunk 22]
```diff
@@ -1,2 +1,49 @@
...
+
+  def not_authorized
+    redirect_to(root_path, alert: I18n.t('auth.errors.not_authorized'))
+  end
+
+  def current_order
+    @current_order ||= obtain_order
+  end
+
+  def obtain_order
```
Comment (yuriy1991) [inline app/controllers/application_controller.rb]: should it be in ApplicationController?

File: app/models/application_record.rb [hunk 23]
```diff
@@ -1,3 +1,8 @@
 class ApplicationRecord < ActiveRecord::Base
   self.abstract_class = true
+
+  MAX_TITLE_LENGTH = 50
```
Comment (yuriy1991) [inline app/models/application_record.rb]: should these constants be in ApplicationRecord?

File: app/queries/best_sellers_query.rb [hunk 24]
```diff
@@ -0,0 +1,17 @@
+class BestSellersQuery
+  def initialize(categories)
+    @categories = categories
+  end
+
+  def call
+    @categories.map do |category|
```
Comment (yuriy1991) [inline app/queries/best_sellers_query.rb]: can we return an AR relation instead of array?

File: bin/rails [hunk 25]
```diff
@@ -1,9 +1,4 @@
 #!/usr/bin/env ruby
-begin
```
Comment (yuriy1991) [inline bin/rails]: why did you delete these lines?

File: config/initializers/active_admin.rb [hunk 26]
```diff
@@ -0,0 +1,315 @@
+ActiveAdmin.setup do |config|
+  # == Site Title
+  #
+  # Set the title that is displayed on the main layout
+  # for each of the active admin pages.
+  #
+  config.site_title = 'Bookstore'
+
+  # Set the link url for the title. For example, to take
```
Comment (yuriy1991) [inline config/initializers/active_admin.rb]: please remove comments

File: config/initializers/devise.rb [hunk 27]
```diff
@@ -0,0 +1,300 @@
...
+  # config.paranoid = true
+
+  # By default Devise will store the user in session. You can skip storage for
+  # particular strategies by setting this option.
+  # Notice that if you are skipping storage for all authentication paths, you
+  # may want to disable generating routes to Devise's sessions controller by
+  # passing skip: :sessions to `devise_for` in your config/routes.rb
+  config.skip_session_storage = [:http_auth]
+
+  # By default, Devise cleans up the CSRF token on authentication to
```
Comment (yuriy1991) [inline config/initializers/devise.rb]: please remove comments

File: app/concepts/book/query/index.rb (hunk 16 above)
Comment (bl0rch1d) [reply app/concepts/book/query/index.rb]: `Find` will throw RecordNotFound exception which will be rescuied with 404 page.

Comment (yuriy1991) [reply app/concepts/book/query/index.rb]: ok

File: app/concepts/address/operation/update.rb (hunk 13 above)
Comment (bl0rch1d) [reply app/concepts/address/operation/update.rb]: image

File: app/controllers/application_controller.rb (hunk 21 above)
Comment (bl0rch1d) [reply app/controllers/application_controller.rb]: Where it can be?

File: app/concepts/coupon/policy/apply.rb (hunk 19 above)
Comment (bl0rch1d) [reply app/concepts/coupon/policy/apply.rb]: User assigns to order only when checkout process started.
Coupon works just like on `https://www.udemy.com/`. It can be applied to anonymous order

File: app/models/application_record.rb (hunk 23 above)
Comment (bl0rch1d) [reply app/models/application_record.rb]: They are used in several forms. How i can refactor this?

File: bin/rails (hunk 25 above)
Comment (bl0rch1d) [reply bin/rails]: I
image

File: app/concepts/address/operation/update.rb (hunk 13 above)
Comment (yuriy1991) [reply app/concepts/address/operation/update.rb]: It won't help:
`false || true` -> true
`false || true || false` -> true

File: app/concepts/coupon/policy/apply.rb (hunk 19 above)
Comment (yuriy1991) [reply app/concepts/coupon/policy/apply.rb]: ok

File: app/controllers/application_controller.rb (hunk 21 above)
Comment (yuriy1991) [reply app/controllers/application_controller.rb]: `expose_address_forms` method is specific for Address, not the whole application
move it to the Controller where it's used

File: app/models/application_record.rb (hunk 23 above)
Comment (yuriy1991) [reply app/models/application_record.rb]: let's move it to model (where it logically belongs). Book?

Comment (bl0rch1d) [reply app/models/application_record.rb]: image

File: app/controllers/application_controller.rb (hunk 21 above)
Comment (bl0rch1d) [reply app/controllers/application_controller.rb]: It using in two controllers:

`AddressesController`
`CheckoutStepsController`

File: app/concepts/address/operation/update.rb [hunk 28]
```diff
@@ -0,0 +1,40 @@
+class Address::Update < Trailblazer::Operation
+  class Present < Trailblazer::Operation
```
Comment (yuriy1991) [inline app/concepts/address/operation/update.rb]: try to move to separate file

File: app/concepts/address/operation/update.rb [hunk 29]
```diff
@@ -0,0 +1,40 @@
...
+
+  step Policy::Guard(Address::Policy::UpdateGuard.new, name: :user), fail_fast: true
+  step Nested(Present)
+  success :extract_params
+  success :define_type
+  step :validate
+  step :persist
+
+  def extract_params(ctx, params:, **)
+    ctx['billing_params'] = params.dig(:user, :billing_address_attributes)
```
Comment (yuriy1991) [inline app/concepts/address/operation/update.rb]: optional: try to define operation based on params in controller

File: app/concepts/book/query/index.rb [hunk 30]
```diff
@@ -0,0 +1,52 @@
+module Book::Query
+  class Index
+    SORTINGS = %w[newest popular high_price low_price title_ascending title_descending].freeze
+
+    def initialize(params)
+      @params = params
+
+      @books = (@params[:category_id] ? Category.find(@params[:category_id]).books : Book.all)
```
Comment (yuriy1991) [inline app/concepts/book/query/index.rb]: please move to method

File: app/concepts/checkout/operation/address.rb [hunk 31]
```diff
@@ -0,0 +1,44 @@
...
+      ctx['shipping_address_form'] = Address::Contract::Create.new(
+        order.shipping_address ||
+        ShippingAddress.new(user.shipping_address&.attributes&.except('id'))
+      )
+    end
+  end
+
+  step Nested(Present)
+  success :extract_params
+  step :validate
```
Comment (yuriy1991) [inline app/concepts/checkout/operation/address.rb]: fast fail?

File: app/concepts/checkout/policy/checkout.rb [hunk 32]
```diff
@@ -0,0 +1,7 @@
+class Checkout::Policy::CheckoutGuard
+  include Uber::Callable
+
+  def call(_ctx, params:, **)
+    params[:current_order].order_items.any?
```
Comment (yuriy1991) [inline app/concepts/checkout/policy/checkout.rb]: please move to controller

File: app/concepts/checkout/policy/step.rb [hunk 33]
```diff
@@ -0,0 +1,30 @@
+class Checkout::Policy::StepGuard
```
Comment (yuriy1991) [inline app/concepts/checkout/policy/step.rb]: try to move this logic to validator

File: app/concepts/order_item/policy/update.rb [hunk 34]
```diff
@@ -0,0 +1,8 @@
+class OrderItem::Policy::UpdateGuard
+  include Uber::Callable
+
+  def call(_ctx, params:, **)
+    params[:current_order].id == params[:order_id].to_i &&
+      params[:current_order].order_items.ids.include?(params[:id].to_i)
```
Comment (yuriy1991) [inline app/concepts/order_item/policy/update.rb]: params[:current_order].order_items.exists?

File: app/controllers/checkout_steps_controller.rb [hunk 35]
```diff
@@ -0,0 +1,59 @@
+class CheckoutStepsController < ApplicationController
+  include Wicked::Wizard
+  include AddressFormsExtractor
+  include CheckoutInitializer
+
+  before_action :fast_authenticate_user!, :initialize_checkout_user
+  steps :address, :shipping, :payment, :confirm, :complete
+
+  def show
+    Checkout::Processor.new(controller: self, action: action_name, step: @step).call
```
Comment (yuriy1991) [inline app/controllers/checkout_steps_controller.rb]: @operation_result = Checkout::Processor.new(controller: self, action: action_name, step: @step).call

File: spec/features/checkout_spec.rb [hunk 36]
```diff
@@ -0,0 +1,253 @@
+describe 'Checkout', type: :feature do
+  let!(:shipping_method) { create(:shipping_method) }
+
+  let(:values) do
```
Comment (yuriy1991) [inline spec/features/checkout_spec.rb]: given

File: spec/features/checkout_spec.rb [hunk 37]
```diff
@@ -0,0 +1,253 @@
...
+      visit order_order_items_path(Order.last)
+    end
+
+    it 'User can buy book through checkout flow' do
+      click_link(I18n.t('checkout.button'))
+
+      expect(page).to have_current_path(checkout_step_path(:address))
+
+      within '.edit_order' do
+        fill_in 'order[billing_address][first_name]',   with: values[:valid][:first_name]
```
Comment (yuriy1991) [inline spec/features/checkout_spec.rb]: each

File: spec/features/user/orders_spec.rb [hunk 38]
```diff
@@ -0,0 +1,151 @@
...
+
+      expect(orders_on_page.size).to eq(user.orders.where(state: I18n.t('order.filter.in_progress')).size)
+    end
+
+    context 'when user uses filter' do
+      it 'see orders with all states' do
+        click_link(class: 'lead')
+        click_link('all')
+
+        sleep 0.5
```
Comment (yuriy1991) [inline spec/features/user/orders_spec.rb]: try to remove sleeps

File: spec/factories/admin_users.rb [hunk 39]
```diff
@@ -0,0 +1,7 @@
+FactoryBot.define do
+  factory :admin_user do
+    email { 'kojima_genius@example.com' }
```
Comment (yaorlov) [inline spec/factories/admin_users.rb]: FFaker?

File: spec/features/admin/order_states_spec.rb [hunk 40]
```diff
@@ -0,0 +1,50 @@
+describe 'Order states' do
+  before do
+    admin_user = create(:admin_user)
+    login_as(admin_user, scope: :admin_user)
+  end
+
+  context 'when admin changes order state' do
+    it 'from in_progress to in_delivery' do
+      create(:order, state: I18n.t('order.filter.in_progress'))
```
Comment (yaorlov) [inline spec/features/admin/order_states_spec.rb]: do not use factory bot in tests, move this logic to `let`

File: spec/features/admin/books_spec.rb [hunk 41]
```diff
@@ -0,0 +1,50 @@
...
+
+      fill_in 'book[height]',       with: book_attributes[:height]
+      fill_in 'book[width]',        with: book_attributes[:width]
+      fill_in 'book[depth]',        with: book_attributes[:depth]
+      fill_in 'book[quantity]',     with: book_attributes[:quantity]
+
+      click_button(I18n.t('admin.book.create'))
+    end
+
+    expect(page).to have_current_path(admin_book_path(Book.last))
```
Comment (yaorlov) [inline spec/features/admin/books_spec.rb]: you should also check that book attributes were saved correctly

File: app/concepts/checkout/addresses/operation/update.rb [hunk 42]
```diff
@@ -0,0 +1,22 @@
...
+
+  def validate(ctx, **)
+    billing = ctx['billing_address_form'].validate(ctx['billing_params'])
+    shipping = ctx['shipping_address_form'].validate(ctx['shipping_params'])
+
+    billing && shipping
+  end
+
+  def persist(ctx, **)
+    ctx['billing_address_form'].save && ctx['shipping_address_form'].save
```
Comment (yaorlov) [inline app/concepts/checkout/addresses/operation/update.rb]: please add ActiveRecord Transaction for saving addresses

File: spec/features/checkout_spec.rb [hunk 43]
```diff
@@ -4,20 +4,26 @@
...
-        last_name: 'Last',
-        address: 'Address Line',
-        city: 'Detroit',
-        zip: '1234',
-        phone: '+380638567656',
-        number: '1234123412341234',
-        card_name: 'Name Of Card',
-        expiration_date: '12/21',
-        cvv: Array.new(rand(3..4)) { rand(1..9) }.join,
+        address: {
```
Comment (yaorlov) [inline spec/features/checkout_spec.rb]: can you use factory bot `attributes` here?

File: db/seeds/book_author.seeds.rb [hunk 44]
```diff
@@ -0,0 +1,33 @@
+SEEDS_COUNT = 50
+BOOK_DIMENSIONS_RANGE = (1.0..10.0).freeze
+
+SEEDS_COUNT.times do |index|
+  Author.create! do |author|
+    author.first_name = FFaker::Name.unique.first_name
+    author.last_name  = FFaker::Name.last_name
+  end
+
+  Book.create! do |book|
```
Comment (yaorlov) [inline db/seeds/book_author.seeds.rb]: You can do it in 1 line `FactoryBot.create_list(:book, 50, with_images: true, images_count: 4)`

File: db/seeds/coupon.seeds.rb [hunk 45]
```diff
@@ -0,0 +1,10 @@
+COUPONS_COUNT = 5
+COUPON_DISCOUNT_RANGE    = (0.01..0.90).freeze
+
+COUPONS_COUNT.times do
+  Coupon.create! do |coupon|
+    coupon.code         = Array.new(6) { rand(65..90).chr }.join
```
Comment (yaorlov) [inline db/seeds/coupon.seeds.rb]: ` Array.new(6) { ('A'..'Z').to_a.sample }.join`

Comment (yaorlov) [reply db/seeds/coupon.seeds.rb]: how do I know what that `chr`s mean?

File: db/seeds/material.seeds.rb [hunk 46]
```diff
@@ -0,0 +1,5 @@
+MATERIALS = ['glossy paper', 'hardcover', 'soft paper', 'cardboard'].freeze
+
+MATERIALS.each do |material|
```
Comment (yaorlov) [inline db/seeds/material.seeds.rb]: the material should be created from the book factory as the relation

File: db/seeds/shipping_method.seeds.rb [hunk 47]
```diff
@@ -0,0 +1,10 @@
+SHIPPING_METHODS_COUNT = 3
+
+SHIPPING_METHODS_COUNT.times do
+  ShippingMethod.create! do |method|
```
Comment (yaorlov) [inline db/seeds/shipping_method.seeds.rb]: `FactoryBot.create_list`

File: spec/concepts/coupon/operation/apply_spec.rb [hunk 48]
```diff
@@ -25,11 +25,12 @@
...
     end
 
-    context 'when coupon has expired' do
-      let(:code) { create(:coupon, :expired).code }
+    context 'when coupon already used by another order' do
+      let(:code) { create(:coupon, used: true).code }
 
       it do
-        expect(result['coupon.relevant']).to be_falsey
+        # binding.pry
```
Comment (yaorlov) [inline spec/concepts/coupon/operation/apply_spec.rb]: remove

**Comment by yaorlov** [comment]
There are some minor bugs:
- Cart counter is not cleared when the order is completed: 
clear_cart_when_order_complete
- I've completed the checkout, but my address was no saved:
order_completed_but_addresses_not_saved
- I followed the placeholder phone pattern, but the phone is still invalid:
phone_validation
- When I apply the shipping method, the shipping price and total are not changed:
shipping_price_and_total

**Comment by yaorlov** [comment]
Please fix phone validation and add https://github.com/rmm5t/strip_attributes for the phone field

File: app/queries/best_sellers_query.rb [hunk 49]
```diff
@@ -0,0 +1,21 @@
...
+  def initialize(categories)
+    @categories = categories
+  end
+
+  def call
+    Book.where(id: bestsellers_ids).order('category_id asc')
+  end
+
+  def bestsellers_ids
+    @categories.map do |category|
```
Comment (yaorlov) [inline app/queries/best_sellers_query.rb]: let's refactor this
use ActiveRecord instead of `map`s
